### PR TITLE
feat(order): add models and model factories

### PR DESCRIPTION
**Pull request by griest024**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Fixes: N/A

## What is the new behavior?
- Adds order models
- Adds order model factories

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

### CODE REVIEW

File: libs/order/package.json [hunk 1]
```diff
@@ -21,5 +21,10 @@
   "peerDependencies": {
     "@angular/core": "0.0.0-PLACEHOLDER",
     "@angular/common": "0.0.0-PLACEHOLDER"
+  },
+  "devDependencies": {
```
Comment (damienwebdev) [inline libs/order/package.json:25]: These are peerDeps as well.

File: libs/order/src/models/order-address.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import { DaffOrderShippingRate } from './order-shipping-rate';
+import { DaffAddress } from '@daffodil/geography';
+
+export interface DaffOrderAddress extends DaffAddress {
```
Comment (damienwebdev) [inline libs/order/src/models/order-address.ts:4]: For now, let's keep this as close to the `DaffAddress` as possible, I think the only thing we really need is `order_id: string | number`.

File: libs/order/src/models/order-item.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import { DaffProductImage } from '@daffodil/product';
+
+export interface DaffOrderItem {
+  item_id: number;
+  image: DaffProductImage;
+  quote_id: number;
```
Comment (damienwebdev) [inline libs/order/src/models/order-item.ts:6]: should be `order_id: string | number`

File: libs/order/src/models/order-item.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  created_at: Date;
+  updated_at: Date;
+  product_id: number;
+  parent_item_id: number;
+  sku: string;
+  name: string;
+  description: string;
```
Comment (damienwebdev) [inline libs/order/src/models/order-item.ts:13]: We don't need description.

File: libs/order/src/models/order-item.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import { DaffProductImage } from '@daffodil/product';
+
+export interface DaffOrderItem {
```
Comment (damienwebdev) [inline libs/order/src/models/order-item.ts:3]: will need:

```
qty_ordered
qty_canceled
qty_fulfilled
```

File: libs/order/src/models/order-payment.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
+export interface DaffOrderPayment {
+  payment_id: number;
+  quote_id: number;
+  created_at: Date;
+  updated_at: Date;
+  method: string; //todo: actually an enum
+  cc_type: string;
+  cc_number_enc: string;
```
Comment (damienwebdev) [inline libs/order/src/models/order-payment.ts:8]: remove this key.

File: libs/order/src/models/order-payment.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  cc_last4: string;
+  cc_cid_enc: string;
```
Comment (damienwebdev) [inline libs/order/src/models/order-payment.ts:10]: remove this key.

File: libs/order/src/models/order-payment.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
+export interface DaffOrderPayment {
```
Comment (damienwebdev) [inline libs/order/src/models/order-payment.ts:1]: This should be a union type of the various possible payment methods., not all payment methods should be forced to a single type.

File: libs/order/src/models/order.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
+import { DaffOrderItem } from './order-item';
+import { DaffOrderAddress } from './order-address';
+import { DaffOrderPayment } from './order-payment';
+
+export interface DaffOrder {
+  id: number;
```
Comment (damienwebdev) [inline libs/order/src/models/order.ts:6]: `id: string | number`

File: libs/order/src/models/order.ts [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+  created_at: Date;
+  updated_at: Date;
+  store_to_base_rate: number;
```
Comment (damienwebdev) [inline libs/order/src/models/order.ts:9]: Remove this key

File: libs/order/src/models/order.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
+import { DaffOrderItem } from './order-item';
+import { DaffOrderAddress } from './order-address';
+import { DaffOrderPayment } from './order-payment';
+
+export interface DaffOrder {
+  id: number;
+  created_at: Date;
```
Comment (damienwebdev) [inline libs/order/src/models/order.ts:7]: string

File: libs/order/src/models/order.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+  updated_at: Date;
```
Comment (damienwebdev) [inline libs/order/src/models/order.ts:8]: string

File: libs/order/src/models/order.ts [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+  store_to_base_rate: number;
+  grand_total: number;
+  checkout_method: string;
```
Comment (damienwebdev) [inline libs/order/src/models/order.ts:11]: remove

File: libs/order/src/models/order.ts [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+  customer_id: number;
+  coupon_code: string;
```
Comment (damienwebdev) [inline libs/order/src/models/order.ts:13]: should be `applied_codes: DaffOrderCoupon[]`

File: libs/order/src/models/order.ts [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+  subtotal: number;
```
Comment (damienwebdev) [inline libs/order/src/models/order.ts:14]: We should probably compress these into a `totals` object.

File: libs/order/src/models/order-address.ts (hunk 2 above)
Comment (griest024) [reply libs/order/src/models/order-address.ts:4]: Should it extend from `DaffPersonalAddress` instead?

Comment (damienwebdev) [reply libs/order/src/models/order-address.ts:4]: Ah, yes. That's a good call.

File: libs/order/src/models/order-item.ts (hunk 3 above)
Comment (griest024) [reply libs/order/src/models/order-item.ts:6]: I'll make the order IDs be `DaffOrder['id']`

File: libs/order/src/models/order-payment.ts (hunk 8 above)
Comment (griest024) [reply libs/order/src/models/order-payment.ts:1]: So we need a type for every payment method? CC, paypal, etc.?

File: libs/order/src/models/order-address.ts (hunk 2 above)
Comment (griest024) [reply libs/order/src/models/order-address.ts]: changed

File: libs/order/src/models/order-item.ts (hunk 5 above)
Comment (griest024) [reply libs/order/src/models/order-item.ts:4]: added

File: libs/order/src/models/order-item.ts (hunk 3 above)
Comment (griest024) [reply libs/order/src/models/order-item.ts]: changed

File: libs/order/src/models/order-item.ts (hunk 4 above)
Comment (griest024) [reply libs/order/src/models/order-item.ts]: removed

File: libs/order/src/models/order-payment.ts (hunk 6 above)
Comment (griest024) [reply libs/order/src/models/order-payment.ts]: removed

File: libs/order/src/models/order-payment.ts (hunk 7 above)
Comment (griest024) [reply libs/order/src/models/order-payment.ts]: removed

File: libs/order/src/models/order.ts (hunk 9 above)
Comment (griest024) [reply libs/order/src/models/order.ts]: changed

File: libs/order/src/models/order.ts (hunk 11 above)
Comment (griest024) [reply libs/order/src/models/order.ts]: changed

File: libs/order/src/models/order.ts (hunk 12 above)
Comment (griest024) [reply libs/order/src/models/order.ts]: changed

File: libs/order/src/models/order.ts (hunk 10 above)
Comment (griest024) [reply libs/order/src/models/order.ts]: removed

File: libs/order/src/models/order.ts (hunk 13 above)
Comment (griest024) [reply libs/order/src/models/order.ts]: removed

File: libs/order/src/models/order.ts (hunk 14 above)
Comment (griest024) [reply libs/order/src/models/order.ts]: changed

File: libs/order/src/models/order.ts (hunk 15 above)
Comment (griest024) [reply libs/order/src/models/order.ts]: changed

**Comment by griest024** [comment]
The factories still need changes as well

**Comment by griest024** [comment]
Factory changes need #873

File: libs/order/src/models/order-invoice.ts [hunk 16]
```diff
@@ -0,0 +1,14 @@
+import { DaffOrderAddress } from './order-address';
+import { DaffOrderPayment } from './order-payment';
+import { DaffOrderShippingMethod } from './order-shipping-method';
+import { DaffOrderShipmentItem } from './order-shipment-item';
+import { DaffOrderTotal } from './order-total';
+
+export interface DaffOrderInvoice {
+  items: DaffOrderShipmentItem[];
```
Comment (damienwebdev) [inline libs/order/src/models/order-invoice.ts:8]: `DaffOrderInvoiceItem`

File: libs/order/src/models/order-payment.ts [hunk 17]
```diff
@@ -0,0 +1,16 @@
+import { DaffOrder } from './order';
+
+export interface DaffOrderCreditCardPayment {
+  payment_id: number;
```
Comment (damienwebdev) [inline libs/order/src/models/order-payment.ts:4]: number | string -- we should probably make an `id` type.

File: libs/order/src/models/order.ts [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+  updated_at: string;
+  status: string;
+  totals: DaffOrderTotal[];
+  applied_codes: DaffOrderCoupon[];
+  items?: DaffOrderItem[];
+  addresses?: DaffOrderAddress[];
+  shipments?: DaffOrderShipment[];
+  payment?: DaffOrderPayment;
+  invoices?: DaffOrderInvoice[];
+  credits?: DaffOrderInvoice[];
```
Comment (damienwebdev) [inline libs/order/src/models/order.ts:22]: I think `credits` should be a rebind of a `DaffOrderInvoice` to a `DaffOrderCredit`

File: libs/order/src/models/order-invoice.ts (hunk 16 above)
Comment (griest024) [reply libs/order/src/models/order-invoice.ts:8]: Is that just an alias or do we need a separate interface entirely?

Comment (damienwebdev) [reply libs/order/src/models/order-invoice.ts:8]: This would just be an alias for now.

File: libs/order/src/models/order.ts (hunk 18 above)
Comment (damienwebdev) [reply libs/order/src/models/order.ts:22]: This too would be an alias.

Comment (griest024) [reply libs/order/src/models/order.ts]: changed

File: libs/order/src/models/order-payment.ts (hunk 17 above)
Comment (griest024) [reply libs/order/src/models/order-payment.ts]: changed

File: libs/order/src/models/order-invoice.ts (hunk 16 above)
Comment (griest024) [reply libs/order/src/models/order-invoice.ts:8]: changed

**Pull request merged**